controllers/secret.go

```diff
@@ -1,37 +1,115 @@
 package controllers
 
 import (
+	"context"
+	"strconv"
+	"strings"
+
 	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 
 	clusterv1alpha1 "github.com/redhat-et/ipfs-operator/api/v1alpha1"
 )
 
+const (
+	peerIDPrefix     = "peerID-"
+	privateKeyPrefix = "privateKey-"
+)
+
+// errorFunc Returns a function which returns the provided
+// error when it is called.
+func errorFunc(err error) controllerutil.MutateFn {
+	return func() error {
+		return err
+	}
+}
+
 func (r *IpfsReconciler) secretConfig(
+	ctx context.Context,
 	m *clusterv1alpha1.Ipfs,
 	sec *corev1.Secret,
 	clusterSecret,
 	bootstrapPrivateKey []byte,
 ) (controllerutil.MutateFn, string) {
 	secName := "ipfs-cluster-" + m.Name
-	expected := &corev1.Secret{
+
+	expectedSecret := &corev1.Secret{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      secName,
 			Namespace: m.Namespace,
 		},
-		Data: map[string][]byte{
-			"CLUSTER_SECRET":          clusterSecret,
-			"BOOTSTRAP_PEER_PRIV_KEY": bootstrapPrivateKey,
-		},
 	}
-	expected.DeepCopyInto(sec)
+	// find secret
+	err := r.Get(ctx, client.ObjectKeyFromObject(expectedSecret), expectedSecret)
+	if err != nil && !errors.IsNotFound(err) {
+		return errorFunc(err), ""
+	}
+	// initialize the secret, if needed
+	if err != nil && errors.IsNotFound(err) {
+		expectedSecret.Data = make(map[string][]byte, 0)
+		expectedSecret.StringData = make(map[string]string, 0)
+		// secret doesn't exist
+		err = generateNewIdentities(expectedSecret, 0, m.Spec.Replicas)
+		if err != nil {
+			return errorFunc(err), ""
+		}
+		expectedSecret.Data["CLUSTER_SECRET"] = clusterSecret
+		expectedSecret.Data["BOOTSTRAP_PEER_PRIV_KEY"] = bootstrapPrivateKey
+	}
+
+	// secret does exist
+	numIdentities := countIdentities(expectedSecret)
+	if numIdentities != m.Spec.Replicas {
+		// create more identities if needed, otherwise they will be reused
+		// when scaling down and then up again
+		if int(numIdentities) < int(m.Spec.Replicas) {
```

Comment:
These `if`s seem redundant. It could `numIdentities < m.Spec.Replicas`. Seems nothing else is done when numIdentities != m.Spec.Replicas.